sections/loans/components/ActionBox/components/Balance.tsx

```diff
@@ -0,0 +1,171 @@
+import { useTranslation } from 'react-i18next';
+
+import synthetix from 'lib/synthetix';
+import { walletAddressState } from 'store/wallet';
+import Connector from 'containers/Connector';
+import { formatUnits } from 'utils/formatters/number';
+import { useLoans } from 'sections/loans/contexts/loans';
+
+type BalanceProps = {
+	asset: string;
+	onSetMaxAmount?: (amount: string) => void;
+};
+
+const Balance: React.FC<BalanceProps> = ({ asset, onSetMaxAmount }) => {
+	const { signer } = Connector.useContainer();
+	return !signer ? null : asset === 'ETH' ? (
+		<ETH {...{ onSetMaxAmount }} />
+	) : (
+		<ERC20 {...{ asset, onSetMaxAmount }} />
+	);
+};
+
+export default Balance;
+
+type ETHProps = {
+	onSetMaxAmount?: (amount: string) => void;
+};
+
+const ETH: React.FC<ETHProps> = ({ onSetMaxAmount }) => {
+	const { t } = useTranslation();
+	const { provider, signer } = Connector.useContainer();
+	const [balance, setBalance] = React.useState(ethers.BigNumber.from('0'));
+
+	const handleSetMaxAmount = () => {
+		if (onSetMaxAmount && balance) {
+			onSetMaxAmount(ethers.utils.formatUnits(balance, 18));
+		}
+	};
+
+	React.useEffect(() => {
+		if (!signer) return;
+
+		let isMounted = true;
+		const unsubs: Array<any> = [() => (isMounted = false)];
+
+		const onSetBalance = async () => {
+			const balance = await signer.getBalance();
+			if (isMounted) setBalance(balance);
+		};
+
+		const subscribe = () => {
+			if (provider) {
+				const newBlockEvent = 'block';
+				provider.on(newBlockEvent, onSetBalance);
+				unsubs.push(() => provider.off(newBlockEvent, onSetBalance));
+			}
+		};
+
+		onSetBalance();
+		subscribe();
+		return () => {
+			unsubs.forEach((unsub) => unsub());
+		};
+	}, [signer, provider]);
+
+	return (
+		balance && (
+			<Container>
+				{t('balance.input-label')} {formatUnits(balance, 18, 2)}{' '}
+				{!onSetMaxAmount ? null : <MaxButton onClick={handleSetMaxAmount} />}
+			</Container>
+		)
+	);
+};
+
+type ERC20Props = {
+	asset: string;
+	onSetMaxAmount?: (amount: string) => void;
+};
+
+const ERC20: React.FC<ERC20Props> = ({ asset, onSetMaxAmount }) => {
+	const { t } = useTranslation();
+	const address = useRecoilValue(walletAddressState);
+	const [balance, setBalance] = React.useState<ethers.BigNumber>(ethers.BigNumber.from('0'));
+	const [decimals, setDecimals] = React.useState<number>(0);
+	const { renBTCContract } = useLoans();
+
+	const handleSetMaxAmount = () => {
+		if (onSetMaxAmount && balance && decimals) {
+			onSetMaxAmount(ethers.utils.formatUnits(balance, decimals));
+		}
+	};
+
+	const contract = React.useMemo(() => {
+		const {
+			contracts: { ProxysBTC: sBTC, ProxysETH: sETH, ProxyERC20sUSD: sUSD },
+		} = synthetix.js!;
+		const tokens: Record<string, ethers.Contract> = { sBTC, sETH, sUSD, renBTC: renBTCContract! };
+		return tokens[asset];
+	}, [asset, renBTCContract]);
+
+	React.useEffect(() => {
+		if (!(contract && address)) return;
+
+		let isMounted = true;
```

Comment:
Not sure I get this. Why are you using `isMounted` exactly?

Reply:
All good @evgenyboxer just explained it to me. Though I wonder if React has a "native" solution for that.

Reply:
Gotcha. There's been talk of a `useAsyncEffect` i think?

Reply:
good call, we could use either that or a custom hook that does the check for us.